File: JavascriptAlgorithm/Array/isValid-020.js
var isValid = function (s) {
  let stack = [];
  let pos = 0;
  for (const c of s) {
    if (c === ")") {
      if (stack[pos - 1] === "(") {
        pos--;
      } else {
        return false;
      }
    } else if (c === "}") {
      if (stack[pos - 1] === "{") {
        pos--;
      } else {
        return false;
      }
    } else if (c === "]") {
      if (stack[pos - 1] === "[") {
        pos--;
      } else {
        return false;
      }
    } else {
      stack[pos++] = c;
    }
  }
  return pos === 0;
};

export { isValid };

File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SecondRound/Chapter04/isValid_020.java
package algo.tzashinorpu.SecondRound.Chapter04;

import java.util.LinkedList;

public class isValid_020 {
    public boolean isValid(String s) {
        LinkedList<Character> stack = new LinkedList<>();
        char[] chars = s.toCharArray();
        for (char c : chars) {
            if (stack.size() == 0) {
                stack.push(c);
            } else {
                if ((c == ')' && stack.peek() != '(') || (c == ']' && stack.peek() != '[') || (c == '}' && stack.peek() != '{')) {
                    return false;
                } else if ((c == ')' && stack.peek() == '(') || (c == ']' && stack.peek() == '[') || (c == '}' && stack.peek() == '{')) {
                    stack.pop();
                } else {
                    stack.push(c);
                }
            }
        }
        return stack.size() == 0;
    }
}
